styleganv2/gan.py

```diff
@@ -0,0 +1,331 @@
+        assert (
+            generator.latent_size == discriminator.latent_size
+            and generator.depth == discriminator.depth
+        ), "Encoder and Decoder are incompatible"
+
+        self.device = device
+        self.generator = generator.to(device)
+        self.discriminator = discriminator.to(device)
+        self.latent_size = generator.latent_size
+        self.input_data_range = input_data_range
+        self.output_data_range = output_data_range
+        self.depth = generator.depth
+        self.latent_distribution = torch.randn
+
+        # wrap the generator and the discriminator networks in DataParallel
+        if self.device != torch.device("cpu"):
+            self.generator = DataParallel(self.generator).to(self.device)
+            self.discriminator = DataParallel(self.discriminator).to(self.device)
+
+        # by default the encoder and the decoder are in eval mode
+        self._toggle_all_networks("eval")
+
+        # create an EMA copy of the generator:
+        self.generator_shadow = copy.deepcopy(self.generator).to(self.device)
+
+        # initialize the gen_shadow weights equal to the
+        # weights of gen
+        update_average(self.generator_shadow, self.generator, beta=0)
+
+    def _toggle_all_networks(self, mode="train"):
+        for network in (self.generator, self.discriminator):
+            if mode.lower() == "train":
+                network.train()
+            elif mode.lower() == "eval":
+                network.eval()
+            else:
+                raise ValueError(f"Unknown mode requested: {mode}")
+
+    def generate_sample(self) -> Tensor:
+        latent = self.latent_distribution((1, self.latent_size)).to(self.device)
+        sample = self.generator([latent for _ in range(self.depth)])
+        return sample.detach().cpu()
+
+    def _save_image_grid(self, images: Tensor, img_file: Path) -> None:
+        normalized_images = adjust_dynamic_range(
+            images.cpu(),
+            drange_in=self.output_data_range,
+            drange_out=(0, 1),
+            slack=True,
+        )
+        torchvision.utils.save_image(
+            normalized_images,
+            img_file,
+            nrow=int(np.ceil(np.sqrt(len(images)))),
+            padding=0,
+        )
+
+    def _r1_penalty(self, real: Tensor) -> Tensor:
+        # forward pass
+        real_scores = self.discriminator(real.requires_grad_(True))
+
+        # perform backward pass from op to merged for obtaining the gradients
+        gradient = torch.autograd.grad(
+            outputs=[torch.sum(real_scores)],
+            inputs=[real],
+            create_graph=True,
+            only_inputs=True,
+            allow_unused=False,
+        )[0]
+
+        # calculate the penalty using these gradients
+        gradient = gradient.view(gradient.shape[0], -1)
+        gradient_penalty = torch.sum(gradient ** 2, dim=-1)
+        return gradient_penalty
+
+    def generator_loss(self, fake: Tensor) -> Tensor:
```

Comment:
Loss functions are not customizable here :disappointed:! 
Perhaps a feature for the next codathon?